Fixes for AutoValue support

msridhar opened the pull request:
1. Don't treat non-`abstract` methods as corresponding to required properties
2. Handle more sophisticated mappings between AutoValue method names and the corresponding Builder method names

msridhar commented:
I can't get the tests to pass here, and I am getting weird errors:

```
7 unexpected diagnostics were found:
AutoValue_GetAnimal.java:95: error: (override.receiver.invalid)
AutoValue_GetAnimal.java:103: error: (override.receiver.invalid)
AutoValue_GetAnimal.java:108: error: (override.receiver.invalid)
AutoValue_GetAnimal.java:113: error: (override.receiver.invalid)
AutoValue_Animal.java:81: error: (override.receiver.invalid)
AutoValue_Animal.java:89: error: (override.receiver.invalid)
AutoValue_Animal.java:94: error: (override.receiver.invalid)
```

The `override.receiver.invalid` errors come from the returns receiver checker, e.g.:
```
/Users/msridhar/git-repos/typesafe-builder-checker/tests/build/testclasses/AutoValue_GetAnimal.java:95: error: [override.receiver.invalid] Incompatible receiver type
    GetAnimal.Builder setName(String name) {
                      ^
  Method
    @This Builder setName(@This Builder this, @MaybeThis String p0) in AutoValue_GetAnimal.Builder
  cannot override
    @MaybeThis Builder setName(@MaybeThis Builder this, @MaybeThis String p0) in GetAnimal.Builder
  found   : @This Builder
  required: @MaybeThis Builder
```
The strange thing is if I delete the new `GetAnimal.java` test file, all the `override.receiver.invalid` errors disappear, including the ones in `AutoValue_Animal.java`, which is an unrelated class!  Also I added a minor variant of `GetAnimal.java` as a test for the returns receiver checker itself and got no `override.receiver.invalid` errors. 

@kelloggm any idea what is going on here?

msridhar commented:
Also if I remove `Animal.java` then the tests pass.  It's only when both `Animal.java` and `GetAnimal.java` are present that the `override.receiver.invalid` errors appear.

kelloggm commented:
I spent some time looking into this. I don't have a solution yet, but I've positively concluded that the problem is in the returns receiver checker, and I have a moderately well-supported hypothesis for what is wrong.

> Also I added a minor variant of GetAnimal.java as a test for the returns receiver checker itself and got no override.receiver.invalid errors.

I was surprised by this claim, and decided to test it for myself. I moved the `GetAnimal.java` file to `returnsrecv-checker/tests/autovalue/`, and got four `override.receiver.invalid` errors:
```
java.lang.AssertionError: The test run was not expected to issue errors/warnings, but it did.
0 out of 0 expected diagnostics were found.
4 unexpected diagnostics were found:
AutoValue_GetAnimal.java:95: error: (override.receiver.invalid)
AutoValue_GetAnimal.java:103: error: (override.receiver.invalid)
AutoValue_GetAnimal.java:108: error: (override.receiver.invalid)
AutoValue_GetAnimal.java:113: error: (override.receiver.invalid)
```
These look like the same errors you were getting from the typesafe builder checker, so that's good. I noticed that the other test (`Animal.java`) in the RR checker omits the *uses* of the builder, so I tried removing those from my copy of `GetAnimal` as well. To my surprise, the errors disappeared. I assume that this is some quirk of AutoValue; perhaps the classes aren't generated if they aren't being used?

After re-adding that code, I tried to trace the source of the error. Reading the long error message, I noticed that the `@This` annotation appears to have been placed correctly in the generated code, but **not in the original code**:
```
    @This Builder setName(@This Builder this, @MaybeThis String p0) in AutoValue_GetAnimal.Builder
  cannot override
    @MaybeThis Builder setName(@MaybeThis Builder this, @MaybeThis String p0) in GetAnimal.Builder
```
To test whether this was the cause of the failure, I manually inserted `@This` annotations in the original code, and the test began to pass. I therefore think that there's something off about the logic for inserting `@This` annotations into builders. We didn't catch it because the `Animal` test in the RR checker apparently isn't actually generating code, which we should fix. I don't know the AutoValue code well enough to diagnose what's wrong with it (though I might have broken it when I added Lombok support). Can you take a look @msridhar and see if there's something wrong there?

msridhar commented:
Thanks for digging! So weird that keeping the uses makes a difference. I
will dig further and figure this out. Thanks, Manu

On Mon, Jul 8, 2019 at 2:50 PM Martin Kellogg <notifications@github.com>
wrote:

> I spent some time looking into this. I don't have a solution yet, but I've
> positively concluded that the problem is in the returns receiver checker,
> and I have a moderately well-supported hypothesis for what is wrong.
>
> Also I added a minor variant of GetAnimal.java as a test for the returns
> receiver checker itself and got no override.receiver.invalid errors.
>
> I was surprised by this claim, and decided to test it for myself. I moved
> the GetAnimal.java file to returnsrecv-checker/tests/autovalue/, and got
> four override.receiver.invalid errors:
>
> java.lang.AssertionError: The test run was not expected to issue errors/warnings, but it did.
> 0 out of 0 expected diagnostics were found.
> 4 unexpected diagnostics were found:
> AutoValue_GetAnimal.java:95: error: (override.receiver.invalid)
> AutoValue_GetAnimal.java:103: error: (override.receiver.invalid)
> AutoValue_GetAnimal.java:108: error: (override.receiver.invalid)
> AutoValue_GetAnimal.java:113: error: (override.receiver.invalid)
>
> These look like the same errors you were getting from the typesafe builder
> checker, so that's good. I noticed that the other test (Animal.java) in
> the RR checker omits the *uses* of the builder, so I tried removing those
> from my copy of GetAnimal as well. To my surprise, the errors
> disappeared. I assume that this is some quirk of AutoValue; perhaps the
> classes aren't generated if they aren't being used?
>
> After re-adding that code, I tried to trace the source of the error.
> Reading the long error message, I noticed that the @This annotation
> appears to have been placed correctly in the generated code, but *not in
> the original code*:
>
>     @This Builder setName(@This Builder this, @MaybeThis String p0) in AutoValue_GetAnimal.Builder
>   cannot override
>     @MaybeThis Builder setName(@MaybeThis Builder this, @MaybeThis String p0) in GetAnimal.Builder
>
> To test whether this was the cause of the failure, I manually inserted
> @This annotations in the original code, and the test began to pass. I
> therefore think that there's something off about the logic for inserting
> @This annotations into builders. We didn't catch it because the Animal
> test in the RR checker apparently isn't actually generating code, which we
> should fix. I don't know the AutoValue code well enough to diagnose what's
> wrong with it (though I might have broken it when I added Lombok support).
> Can you take a look @msridhar <https://github.com/msridhar> and see if
> there's something wrong there?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kelloggm/typesafe-builder-checker/pull/21?email_source=notifications&email_token=AABPEUMPXTPGEV3WGT2OMULP6OZAXA5CNFSM4H6XAT4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZOOR3Q#issuecomment-509405422>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABPEUMMZHWBEEWEG6NJP5DP6OZAXANCNFSM4H6XAT4A>
> .
>

msridhar commented:
When rebased on #24 and with the latest returns-receiver fixes, the tests pass locally for me now for this PR.  Will update the PR once #24 lands

kelloggm commented on src/main/java/org/checkerframework/checker/builder/TypesafeBuilderAnnotatedTypeFactory.java:
```diff
@@ -465,6 +477,27 @@ private static boolean hasAnnotation(
+  private static String autoValuePropToBuilderSetterName(
+      String prop, Set<String> allBuilderMethodNames) {
+    // property name may be prefixed JavaBean style with 'get' or 'is'; strip that part if present
+    if (prop.startsWith("get") && Character.isUpperCase(prop.charAt(3))) {
```
Is it possible for the entire name of the method to be `get()`? For example, what if I named the property "get". Would it be possible for me to direct AutoValue to produce a method that doesn't have a prefix and then cause this code to throw a `StringOutOfBoundsException`

msridhar replied:
Good catch!  I added a fix and a test
